Sources/NIO/IntegerBitPacking.swift

```diff
@@ -0,0 +1,117 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@usableFromInline
+enum _IntegerBitPacking {}
+
+extension _IntegerBitPacking {
+    @inlinable
+    static func packUU<Left: FixedWidthInteger & UnsignedInteger,
+                       Right: FixedWidthInteger & UnsignedInteger,
+                       Result: FixedWidthInteger & UnsignedInteger>(_ l: Left,
+                                                                    _ r: Right,
+                                                                    type: Result.Type = Result.self) -> Result {
+        assert(MemoryLayout<Left>.size + MemoryLayout<Right>.size <= MemoryLayout<Result>.size)
+        assert(Left.max.nonzeroBitCount == Left.bitWidth)
```

Comment:
What's the purpose of this assertion?

Reply:
@Lukasa I'm relying on `.max` being usable to mask the value. So if you were to create a weird integer type where that isn't true, then you couldn't use this. None of the types exist in the Swift stdlib (or in NIO) but I thought better be safe :P

Reply:
I think the protocol forbids the kind of weirdness you're defending against. The documentation on `FixedWidthInteger.max` says:

> For unsigned integer types, this value is `(2 ** bitWidth) - 1`, where `**` is exponentiation. For signed integer types, this value is `(2 ** (bitWidth - 1)) - 1`.

Reply:
@Lukasa thank you! Want me to kill the assertion or leave it in just in case somebody makes such a type?

Reply:
I'm inclined to kill it entirely. This isn't API and we're unlikely to make it API so I don't think the risk is very high at all.

Reply:
done